code/modules/multiz/level_data.dm

```diff
@@ -224,6 +233,111 @@
 	if(!skip_gen)
 		generate_level()
 	after_generate_level()
+
+	// Determine our relative positioning.
+	// First find an appropriate origin point.
+	var/datum/level_data/origin
+	for(var/check_z in SSmapping.get_connected_levels(level_z, include_lateral = TRUE))
+		var/datum/level_data/checking = SSmapping.levels_by_z[check_z]
+		if(!isnull(checking.z_volume_level_x) && !isnull(checking.z_volume_level_y) && !isnull(checking.z_volume_level_z))
+			origin = checking
+			break
+
+	// If nobody else has set up lateral level coords, we must be the origin.
+	if(!origin)
+		origin = src
+		z_volume_level_x = 0
+		z_volume_level_y = 0
+		z_volume_level_z = 0
+
+	var/list/checked  = list()
+	var/list/to_check = list(origin)
+	while(length(to_check))
+
+		// Update our tracking lists.
+		var/datum/level_data/checking = to_check[1]
+		var/datum/level_data/previous = to_check[checking]
+		to_check -= checking
+		checked |= checking
+
+		// Obtain all our neighbors for flood fill.
+		for(var/level_id in checking.connected_levels)
+			var/datum/level_data/neighbor = SSmapping.levels_by_id[level_id]
+			if(istype(neighbor) && !(neighbor in checked))
+				to_check[neighbor] = checking
+		if(HasBelow(checking.level_z))
+			var/datum/level_data/neighbor = SSmapping.levels_by_z[checking.level_z-1]
+			if(istype(neighbor) && !(neighbor in checked))
+				to_check[neighbor] = checking
+		if(HasAbove(checking.level_z))
+			var/datum/level_data/neighbor = SSmapping.levels_by_z[checking.level_z+1]
+			if(istype(neighbor) && !(neighbor in checked))
+				to_check[neighbor] = checking
+
+		// Update our lateral coords based on the direction of our connection.
+		// This does NOT appropriately handle looping or non-horizontal/vertical
+		// connections, which violate our assumptions.
+		if(previous)
+
+			checking.z_volume_level_x = previous.z_volume_level_x
+			checking.z_volume_level_y = previous.z_volume_level_y
+			checking.z_volume_level_z = previous.z_volume_level_z
+
+			var/connect_dir = checking.connected_levels[previous.level_id]
```

Comment:
previous can be non-null even when `checking.connected_levels` is null which causes a bad index runtime here. should probably do uh
```suggestion
			var/connect_dir = checking.connected_levels?[previous.level_id]
```
or use lazyaccess, but in this case they're equivalent